client/my-sites/plan-features-2023-grid/index.tsx

```diff
@@ -1180,6 +1148,47 @@ const ConnectedPlanFeatures2023Grid = connect(
 
 		const isJetpackNotAtomic = isJetpack && ! isSiteAT;
 
+		const hasDiscountNotice = () => {
+			const { withDiscount } = ownProps;
+			const bannerContainer = getBannerContainer();
+			if ( ! bannerContainer ) {
+				return false;
+			}
+
+			const activeDiscount = getDiscountByName( withDiscount );
+			if ( ! activeDiscount || hasPlaceholders( planProperties ) || ! canUserPurchasePlan ) {
+				return false;
+			}
+			return true;
+		};
+
+		const higherPlanAvailable = () => {
+			const currentPlan = currentSitePlan?.productSlug;
+			const highestPurchasablePlan = planProperties
+				.filter( ( { planName } ) => planName !== PLAN_ENTERPRISE_GRID_WPCOM )
+				.filter( ( { availableForPurchase } ) => availableForPurchase )
+				.pop();
+			const { planName: highestPurchasablePlanName } = highestPurchasablePlan ?? {};
+			return currentPlan !== highestPurchasablePlanName;
+		};
+
+		const isDiscountNoticeVisible = hasDiscountNotice();
+		const isFreePlanActive = isFreePlan( sitePlan?.product_slug ?? '' );
+		const isWPPlansPage = sitePlan && ! isJetpackNotAtomic && ! isInSignup;
+		const isHigherPlanAvailable = higherPlanAvailable();
+		const bannerContainer = getBannerContainer();
+
+		const hasPlaceholdersFlag = hasPlaceholders( planProperties );
+		const hideCreditNotice =
+			! isWPPlansPage ||
+			isFreePlanActive ||
+			isDiscountNoticeVisible ||
+			hasPlaceholdersFlag ||
+			! canUserPurchasePlan ||
+			! bannerContainer ||
+			! planCredits ||
+			! isHigherPlanAvailable;
```

Comment:
This is quite a long expression that I'd rather if it lives somewhere else or that we break apart. the interplay between "hiding the credit notice" (`hideCreditNotice`) and "the discount notice being visible" (`isDiscountNoticeVisible`) is also hidden within, and we are also passing these along as separate props. 

I don't have solutions, but it's hard to digest all this logic. Maybe see if anything can be cleaned up and we can re-review. 🙇

Reply:
This is actually cleaning up a loosely defined set of conditions from before.
The initial set of flags were taken from below.

https://github.com/Automattic/wp-calypso/blob/2bd7e469b071c51d3ddc0f69805439e0b29e795f/client/my-sites/plan-features-2023-grid/index.tsx#L1003-L1012

The `isShowPlanCreditsApplied` above further expands into another set of flags from below

https://github.com/Automattic/wp-calypso/blob/2bd7e469b071c51d3ddc0f69805439e0b29e795f/client/my-sites/plan-features-2023-grid/index.tsx#L1003-L1012

I do see a cleanup for this which involves moving all of this notice logic into it's own component. Some complexity also arises from using `React.portal` where we are directly mutating the DOM. Cleaning this up can potentially eliminate `bannerContainer` which does a DOM check to see whether the mounting element exists. 
https://github.com/Automattic/wp-calypso/blob/2bd7e469b071c51d3ddc0f69805439e0b29e795f/client/my-sites/plan-features-2023-grid/index.tsx#L874-L876

I was initially worried that tampering too much will break something unrelated. But looks like this is a pretty isolated change done at #72952 and the code volume to be manageable. Let me take another stab at this refactor!